core/math/math_funcs.h

```diff
@@ -74,11 +74,23 @@ class Math {
 	static _ALWAYS_INLINE_ double tanh(double p_x) { return ::tanh(p_x); }
 	static _ALWAYS_INLINE_ float tanh(float p_x) { return ::tanhf(p_x); }
 
-	static _ALWAYS_INLINE_ double asin(double p_x) { return ::asin(p_x); }
-	static _ALWAYS_INLINE_ float asin(float p_x) { return ::asinf(p_x); }
+	static _ALWAYS_INLINE_ double safe_asin(double p_x) { return p_x < -1 ? (-Math_PI / 2) : (p_x > 1 ? (Math_PI / 2) : ::asin(p_x)); }
+	static _ALWAYS_INLINE_ float safe_asin(float p_x) { return p_x < -1 ? (-Math_PI / 2) : (p_x > 1 ? (Math_PI / 2) : ::asinf(p_x)); }
 
-	static _ALWAYS_INLINE_ double acos(double p_x) { return ::acos(p_x); }
-	static _ALWAYS_INLINE_ float acos(float p_x) { return ::acosf(p_x); }
+	static _ALWAYS_INLINE_ double unsafe_asin(double p_x) { return ::asin(p_x); }
+	static _ALWAYS_INLINE_ float unsafe_asin(float p_x) { return ::asinf(p_x); }
+
+	static _ALWAYS_INLINE_ double asin(double p_x) { return safe_asin(p_x); }
+	static _ALWAYS_INLINE_ float asin(float p_x) { return safe_asin(p_x); }
```

Comment:
I understand the intent to have both an explicit `safe_asin` and the main `asin` which is safe by default, but having two synonymous methods just means that we'll introduce inconsistency across the codebase where some contributors use `asin` and others use `safe_asin`.

Reply:
This is a tricky one. I added the explicit `safe_acos` to replace parts of code that were already clamping. This signifies areas where the author / bugfix has shown clamping to be absolutely necessary.

In general there are three cases rather than two:
1) You absolutely know you must clamp (`safe_acos`)
2) You aren't sure, and not a bottleneck, so might as well (`acos`)
3) You have determined the Nan condition can't happen and / or this is a bottleneck and Nans are dealt with (`unsafe_acos`)

We could alternatively use a comment to specify that a clamp is necessary instead of `safe_acos` (1), but having all three is clear and easily searchable.

I can adjust the PR to use comments to distinguish (1) and (2) but imo I would caution against it. What happens when you later come to optimize a function that uses `acos`? You now no longer know whether it must be safe or not (unless you re-analyse the code, which may be complex for math) - has someone missed a comment etc.

I can add some comment explanation to the `math_funcs.h` file to make it clearer when to use each version, as a guide for contributors. :thinking:

Reply:
I think this might benefit from some deeper analysis of the uses of the trigonometric functions, from the case in #8111 there might be reasons to use `atan2` instead in situations where it is possible, i.e. where a well defined x and y can be found, I think fixing the immediate problem of NaNs might be sufficient to merge this with a placeholder function and then investigate individual cases to follow up

Reply:
(Not sure about real code) But at least in micro benchmarks the `unsafe` version is only about 0.5% faster then the branched version (if all inputs are actually in the -1 - 1 range, otherwise the branched version quickly wins), so I am not sure if the unsafe version of these functions is actually necessary at all.

```
----------------------------------------------------------
Benchmark                Time             CPU   Iterations
----------------------------------------------------------
BenchAsinClamp       30815 ns        30815 ns        22743
BenchAsinBranch      27037 ns        27037 ns        25850
BenchAsinUnsafe      26934 ns        26934 ns        26773
```

Reply:
@RedworkDE that's actually a good point. The unsafe version by definition shouldn't be hitting the -1 / +1 limit. I guess we could do this. The other option is that the unsafe version is so rare (not even used as yet) that we could directly call `::acos` at the call site with a comment. :thinking:

Reply:
@RedworkDE comparing unsafe to the branching, I'm getting unsafe being around:
4% faster (when some are outside range)
10% faster (when all inside range)

This probably depends very much on the benchmark. But in general the clamped version shouldn't be _that_ much slower (and in practice, often cache is the bottleneck rather than these instructions).